mkrlwe/marshaler.go

```diff
@@ -0,0 +1,332 @@
+	pubKey.Key[1] = new(MKDecomposedPoly)
+
+	separator := binary.LittleEndian.Uint64(data[:8])
+	err1 := pubKey.Key[0].UnmarshalBinary(data[8 : 8+separator])
+	err2 := pubKey.Key[1].UnmarshalBinary(data[8+separator : len(data)-8])
+
+	pubKey.PeerID = binary.LittleEndian.Uint64(data[len(data)-8:])
+
+	if err1 != nil || err2 != nil {
+		panic("Couldn't UnMarshall evaluation key")
+	}
+
+	return nil
+}
+
+// MarshalBinary encodes a Galois evaluation key in a byte slice.
+func (evalGal *MKRotationKey) MarshalBinary() (data []byte, err error) {
+
+	d1, err1 := MarshalSwitchingKey(evalGal.Key)
+	d2 := make([]byte, 8)
+	binary.LittleEndian.PutUint64(d2, evalGal.PeerID)
+
+	if err1 != nil {
+		panic("Couldn't Marshall evaluation key")
+	}
+	data = make([]byte, len(d1)+8)
+	for i := 0; i < len(d1); i++ {
+		data[i] = d1[i]
+	}
+	for i := 0; i < len(d2); i++ {
+		data[len(d1)+i] = d2[i]
+	}
+
+	return data, nil
+}
+
+// UnmarshalBinary decodes a previously marshaled Galois Evaluation Key in the target MKEvalGalKey.
+func (evalGal *MKRotationKey) UnmarshalBinary(data []byte) (err error) {
+
+	evalKey, err := UnmarshalSwitchingKey(data[:len(data)-8])
+	pid := binary.LittleEndian.Uint64(data[len(data)-8:])
+
+	if err != nil {
+		panic("Couldn't UnMarshall evaluation key")
+	}
+	evalGal.Key = &evalKey
+	evalGal.PeerID = pid
+
+	return nil
+}
+
+// MarshalBinary encodes an evaluation key in a byte slice.
+func (evalKey *MKRelinearizationKey) MarshalBinary() (data []byte, err error) {
+
+	// marshal each component of evalKey
+	d1, err1 := MarshalSwitchingKey(evalKey.Key01)
+	d2, err2 := evalKey.Key2.MarshalBinary()
+	d3 := make([]byte, 8)
+	binary.LittleEndian.PutUint64(d3, evalKey.PeerID)
+	if err1 != nil || err2 != nil {
+		panic("Couldn't Marshall evaluation key")
+	}
+
+	// Merge the byte slices
+	d12 := MergeTwoByteSlices(d1, d2)
+	data = make([]byte, len(d12)+8)
+	for i := 0; i < len(d12); i++ {
+		data[i] = d12[i]
+	}
+	for i := 0; i < len(d3); i++ {
+		data[len(d12)+i] = d3[i]
+	}
+
+	return data, nil
+}
+
+// UnmarshalBinary decodes a previously marshaled evaluation key in the target MKEvaluationKey.
+func (evalKey *MKRelinearizationKey) UnmarshalBinary(data []byte) (err error) {
+
+	// Unmarshal the switchingKey first
+	separator := binary.LittleEndian.Uint64(data[:8])
+	swkSize := binary.LittleEndian.Uint64(data[8:16])
+	value := make([][2]*ring.Poly, swkSize/2)
+
+	swk := new(rlwe.SwitchingKey)
+	var pointer, inc uint64
+	pointer = 16
+	dpStartIndex := separator + 8
+
+	for j := uint64(0); j < swkSize/2; j++ {
+		for k := 0; k < 2; k++ {
+			value[j][k] = new(ring.Poly)
+			if inc, err = value[j][k].DecodePolyNew(data[pointer:dpStartIndex]); err != nil {
+				return err
+			}
+			pointer += inc
+
+		}
+	}
+	swk.Value = value
+	evalKey.Key01 = swk
+
+	// Unmarshal the decomposedPoly
+	k2 := new(MKDecomposedPoly)
+	dpSize := binary.LittleEndian.Uint64(data[dpStartIndex : dpStartIndex+8])
+	dpoly := make([]*ring.Poly, dpSize)
+	pointer = dpStartIndex
```

Comment:
Hi. Marshalling and un-marshalling a mk relinearization key produces errors for the field `Key2`. I think the pointer here should take value `dpStartIndex + 8`, so it can omit the bytes that indicate the length of the encoded string.